torch/optim/radam.py

```diff
@@ -115,14 +121,140 @@ def step(self, closure=None):
+                         weight_decay: float,
+                         eps: float):
+
+    for i, param in enumerate(params):
+        grad = grads[i]
+        exp_avg = exp_avgs[i]
+        exp_avg_sq = exp_avg_sqs[i]
+        step = state_steps[i]
+
+        bias_correction1 = 1 - beta1 ** step
+        bias_correction2 = 1 - beta2 ** step
+
+        if weight_decay != 0:
+            grad = grad.add(param, alpha=weight_decay)
+
+        # Decay the first and second moment running average coefficient
+        exp_avg.mul_(beta1).add_(grad, alpha=1 - beta1)
+        exp_avg_sq.mul_(beta2).addcmul_(grad, grad, value=1 - beta2)
+
+        # correcting bias for the first moving moment
+        bias_corrected_exp_avg = exp_avg / bias_correction1
+
+        # maximum length of the approximated SMA
+        rho_inf = 2 / (1 - beta2) - 1
+        # compute the length of the approximated SMA
+        rho_t = rho_inf - 2 * step * (beta2 ** step) / bias_correction2
+
+        if rho_t > 5.:
+            # Compute the variance rectification term and update parameters accordingly
+            rect = math.sqrt((rho_t - 4) * (rho_t - 2) * rho_inf / ((rho_inf - 4) * (rho_inf - 2) * rho_t))
+            adaptive_lr = math.sqrt(bias_correction2) / exp_avg_sq.sqrt().add_(eps)
+
+            param.add_(bias_corrected_exp_avg * lr * adaptive_lr * rect, alpha=-1.0)
+        else:
+            param.add_(bias_corrected_exp_avg * lr, alpha=-1.0)
+
+
+def _multi_tensor_radam(params: List[Tensor],
+                        grads: List[Tensor],
+                        exp_avgs: List[Tensor],
+                        exp_avg_sqs: List[Tensor],
+                        state_steps: List[int],
+                        *,
+                        beta1: float,
+                        beta2: float,
+                        lr: float,
+                        weight_decay: float,
+                        eps: float):
+
+    # maximum length of the approximated SMA
+    rho_inf = 2 / (1 - beta2) - 1
+    # compute the length of the approximated SMA
+    rho_t_list = [rho_inf - 2 * step * (beta2 ** step) / (1 - beta2 ** step) for step in state_steps]
+
+    bias_correction1 = [1 - beta1 ** step for step in state_steps]
+    bias_correction2 = [1 - beta2 ** step for step in state_steps]
+    if weight_decay != 0:
+        torch._foreach_add_(grads, params, alpha=weight_decay)
+
+    # Decay the first and second moment running average coefficient
+    torch._foreach_mul_(exp_avgs, beta1)
+    torch._foreach_add_(exp_avgs, grads, alpha=1 - beta1)
+
+    torch._foreach_mul_(exp_avg_sqs, beta2)
+    torch._foreach_addcmul_(exp_avg_sqs, grads, grads, 1 - beta2)
+
+    rect = [math.sqrt((rho_t - 4) * (rho_t - 2) * rho_inf / ((rho_inf - 4) * (rho_inf - 2) * rho_t))
+            if rho_t > 5 else 0 for rho_t in rho_t_list]
+    unrectified = [0 if rect > 0 else 1. for rect in rect]
+
+    exp_avg_sq_sqrt = torch._foreach_add(torch._foreach_sqrt(exp_avg_sqs), eps)
```

Comment:
`eps` was unused in the multitensor radam implementation but used in the single tensor implentation, I believe the correct place to add it is here but I am not super sure whether this omission was intentional, should i be adding this here?

Reply:
I wouldn't change the foreach implementation at all in this PR and open an issue about this.